controllers/user.go

```diff
@@ -0,0 +1,123 @@
+package controllers
+
+import (
+	"encoding/json"
+	"net/http"
+	"regexp"
+	"strconv"
+
+	"github.com/lifeiscontent/webservice/models"
+)
+
+type userController struct {
+	userIDPattern *regexp.Regexp
+}
+
+func (uc userController) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path == "/users" {
+		switch r.Method {
+		case http.MethodGet:
+			uc.getAll(w, r)
+		case http.MethodPost:
+			uc.post(w, r)
+		default:
+			w.WriteHeader(http.StatusNotImplemented)
+		}
+	} else {
+		matches := uc.userIDPattern.FindStringSubmatch(r.URL.Path)
+		if len(matches) == 0 {
+			w.WriteHeader(http.StatusNotFound)
+		}
+		id, err := strconv.Atoi(matches[1])
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+		}
+		switch r.Method {
+		case http.MethodGet:
+			uc.get(id, w)
+		case http.MethodPut:
+			uc.put(id, w, r)
+		case http.MethodDelete:
+			uc.delete(id, w)
+		default:
+			w.WriteHeader(http.StatusNotImplemented)
+		}
+	}
+}
+
+func (uc *userController) getAll(w http.ResponseWriter, r *http.Request) {
+	encodeResponseAsJSON(models.GetUsers(), w)
+}
+
+func (uc *userController) get(id int, w http.ResponseWriter) {
+	u, err := models.GetUserByID(id)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	encodeResponseAsJSON(u, w)
+}
+
+func (uc *userController) post(w http.ResponseWriter, r *http.Request) {
+	u, err := uc.parseRequest(r)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("Could not parse User object"))
+		return
+	}
+	u, err = models.AddUser(u)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+	encodeResponseAsJSON(u, w)
+}
+
+func (uc *userController) put(id int, w http.ResponseWriter, r *http.Request) {
+	u, err := uc.parseRequest(r)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("Could not parse User object"))
+		return
+	}
+	if id != u.ID {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("ID of submitted user must match ID in URL"))
+		return
+	}
+	u, err = models.UpdateUser(u)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+	encodeResponseAsJSON(u, w)
+}
+
+func (uc *userController) delete(id int, w http.ResponseWriter) {
+	err := models.RemoveUserById(id)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (uc *userController) parseRequest(r *http.Request) (models.User, error) {
+	dec := json.NewDecoder(r.Body)
+	var u models.User
+	err := dec.Decode(&u)
+	if err != nil {
+		return models.User{}, err
+	}
+	return u, nil
+}
+
+func newUserController() *userController {
```

Comment:
@zigara what does this line mean to you?

I see a function which returns a data structure of `userController` but I don't think that is correct. Can you explain?

Reply:
This is a function that returns a pointer to a userController instance.